rfcs/0013-changes-lifecyle.md

```diff
@@ -0,0 +1,349 @@
+- Start Date: 2024-08-05
+- RFC PR: [FuelLabs/sway-rfcs#41](https://github.com/FuelLabs/sway-rfcs/pull/41)
+
+# Summary
+
+[summary]: #summary
+
+This RFC is a guide on how to introduce changes following the compiler release lifecycle.
+
+# Motivation
+
+[motivation]: #motivation
+
+To maximize user experience when updating the compiler, we need a guide on how to introduce changes.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The compiler will follow a "rolling release" scheme, which means that periodically (to be specified) a
+new major version will be released.
+
+This means that as soon as the compiler reaches version "1.0.0", the next **version** will be "1.1.0"; and
+after that and if needed, the next **patch** would be "1.1.1".
+
+The only changes that will trigger "patches" are:
+
+1. Urgent security fixes;
+2. Fixing bugs that render "stable" functionality unusable;
+
+## Release Notes
+
+To track all release notes, a special file named `ReleaseNotes.md` will be created in the repository's root directory.
+
+When a version is being launched, for example, `1.10` a new section `1.11-nightly` will be created empty. From this time on, all new PRs will create a new item under the "next version", `1.11-nightly` in this example.
+
+Each item will follow the template:
+
+```
+# Version 1.11-nightly
+
+## Forc
+
+### Added
+### Changed
+### Deprecated
+### Removed
+### Fixed
+### Security
+
+## Tools
+
+...
+
+## Sway
+
+...
+
+## Standard Libraries
+
+...
+
+# Version 1.10
+
+...
+```
+
+Inside each section, items will follow the template of a user friendly one line description nad link to the PR.
+
+```
+- Index operator using Index trait [#6356](https://github.com/FuelLabs/sway/pull/6356)
+```
+
+Given that all PRs will touch this file, to avoid conflicts and decrease the experience when merging them, we will use "git custom merge driver" (see https://git-scm.com/docs/gitattributes#_defining_a_custom_merge_driver).
+
+This allows a custom merge strategy to a specific file using ".gitattributes".
+
+```
+ReleaseNotes.md merge=union
+```
+
+## Breaking changes
+
+A breaking change is a difference in functionality from the previous version of the compiler that may require an update to sway code in order for it to compile.
+
+The following changes are defined to be breaking changes and will need to follow the process of being gated by feature flags.
+
+1. Code without deprecated warning to be flagged as an error;
+2. ABI JSON properties to be removed or have their type changed;
+3. Binary encoding that will break how SDK (and others) communicate with the compiler, including:
+   - How scripts/predicates accept arguments on `main`;
+   - How scripts/predicates return data from `main`;
+   - How the `contract method selector` is encoded;
+   - How `contract method` arguments are encoded;
+   - How `log` data is encoded;
+   - How `message` data is encoded;
+4. Utilization of new VM opcodes;
```

Comment:
I think this should be clarified.

Fully additive changes to the opcode set (say by reinterpreting reserved opcodes) should not be considered a breaking change in my opinion.

Only that which changes or breaks existing user code should be breaking in the ideal, new vm additions that purely extend the feature set do not do this.

Reply:
My idea was similar to IR changes. Tools that read IR/binary opcodes will break with new opcodes. That is why these would need to be introduced as breaking changes.

Reply:
Those tools support a particular version of the VM. So long as we produce legal bytecode for that particular VM, I don't think that's a breaking change, even if some are new.

And VM changes are already a breaking change. So yeah I don't think we want it.